## Peer-graded Assignment: Segmenting and Clustering Neighborhoods in Toronto

# Problem - 1

In [3]:
import pandas as pd
import numpy as np

In [4]:
import urllib.request

In [5]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"

In [6]:
page = urllib.request.urlopen(url)

In [8]:
from bs4 import BeautifulSoup

In [10]:
soup = BeautifulSoup(page, "lxml")

In [87]:
#print(soup.prettify())

In [11]:
soup.title.string

'List of postal codes of Canada: M - Wikipedia'

In [12]:
all_tables=soup.find_all("table")

In [88]:
#all_tables

In [13]:
right_table=soup.find('table', class_='wikitable sortable')
#right_table

In [14]:
A=[]
B=[]
C=[]


for row in right_table.findAll('tr'):
    cells=row.findAll('td')
    if len(cells)==3:
        A.append(cells[0].find(text=True))
        B.append(cells[1].find(text=True))
        C.append(cells[2].find(text=True))

In [43]:
import pandas as pd
df=pd.DataFrame(A,columns=['Postal Code'])
df['Borough']=B
df['Neighbourhood']=C
df.shape
df.head()

Postal Code            Borough               Neighbourhood
0        M1A
      Not assigned
               Not assigned

1        M2A
      Not assigned
               Not assigned

2        M3A
        North York
                  Parkwoods

3        M4A
        North York
           Victoria Village

4        M5A
  Downtown Toronto
  Regent Park, Harbourfront

In [16]:
df[(df['Borough'] == "Not assigned\n")].count()

Postal Code      77
Borough          77
Neighbourhood    77
dtype: int64

In [44]:
df = df[df.Borough != "Not assigned\n"]

In [45]:
df.head()

Postal Code            Borough                                 Neighbourhood
2        M3A
        North York
                                    Parkwoods

3        M4A
        North York
                             Victoria Village

4        M5A
  Downtown Toronto
                    Regent Park, Harbourfront

5        M6A
        North York
             Lawrence Manor, Lawrence Heights

6        M7A
  Downtown Toronto
  Queen's Park, Ontario Provincial Government

In [19]:
df.shape

(103, 3)

In [90]:
# Checking for duplicates in Postal Code
#df.groupby('Postal Code').count() > 1
# No duplciates found

In [20]:
df[(df['Neighbourhood'] == "Not assigned\n")].count()
# All neighborhoods have values

Postal Code      0
Borough          0
Neighbourhood    0
dtype: int64

In [21]:
df.shape

(103, 3)

## Peer-graded Assignment: Segmenting and Clustering Neighborhoods in Toronto¶
# Problem - 2

In [92]:
!conda install -c conda-forge geocoder -y

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geocoder


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    ca-certificates-2020.6.20  |       hecda079_0         145 KB  conda-forge
    certifi-2020.6.20          |   py36h9f0ad1d_0         151 KB  conda-forge
    ratelim-0.1.6              |             py_2           6 KB  conda-forge
    geocoder-1.38.1            |             py_1          53 KB  conda-forge
    openssl-1.1.1g             |       h516909a_0         2.1 MB  conda-forge
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.5 MB

The following NEW packages will be INSTALLED:

    geocoder:        1.38.1-py_1       conda-forge
    python_abi:    

In [ ]:
import geocoder # import geocoder

# initialize your variable to None
lat_lng_coords = None

# loop until you get the coordinates
while(lat_lng_coords is None):
  g = geocoder.google('{}, Toronto, Ontario'.format('Borough'))
  lat_lng_coords = g.latlng

latitude = lat_lng_coords[0]
longitude = lat_lng_coords[1]


In [ ]:
# Geocoder is not working, using the csv file provided in assignement

In [46]:
df_geo=pd.read_csv("http://cocl.us/Geospatial_data")

In [47]:
df_geo.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [48]:
df_geo.shape

(103, 3)

In [173]:
df_geo.columns

Index(['Postal Code', 'Latitude', 'Longitude'], dtype='object')

In [174]:
df.columns

Index(['Postal Code', 'Borough', 'Neighbourhood'], dtype='object')

In [49]:
df.columns = df.columns.str.replace(' ', '')

In [50]:
df_geo.columns = df_geo.columns.str.replace(' ', '')

In [51]:
df.rename(columns={'Postal Code':'Postal_Cd'}, inplace=True)
df.columns

Index(['PostalCode', 'Borough', 'Neighbourhood'], dtype='object')

In [52]:
df_geo.rename(columns={'Postal Code':'Postal_Cd'}, inplace=True)
df_geo.columns

Index(['PostalCode', 'Latitude', 'Longitude'], dtype='object')

In [53]:
df['PostalCode'] = df['PostalCode'].str.replace('\n','')

In [56]:
t=pd.merge(df, df_geo, on='PostalCode',how='inner')
# This is second problem datafarme
t

PostalCode            Borough  \
0          M3A        North York
   
1          M4A        North York
   
2          M5A  Downtown Toronto
   
3          M6A        North York
   
4          M7A  Downtown Toronto
   
5          M9A         Etobicoke
   
6          M1B       Scarborough
   
7          M3B        North York
   
8          M4B         East York
   
9          M5B  Downtown Toronto
   
10         M6B        North York
   
11         M9B         Etobicoke
   
12         M1C       Scarborough
   
13         M3C        North York
   
14         M4C         East York
   
15         M5C  Downtown Toronto
   
16         M6C              York
   
17         M9C         Etobicoke
   
18         M1E       Scarborough
   
19         M4E      East Toronto
   
20         M5E  Downtown Toronto
   
21         M6E              York
   
22         M1G       Scarborough
   
23         M4G         East York
   
24         M5G  Downtown Toronto
   
25         M6G  Downtown Toronto
   
26         M1H       Scarborough
   
27         M2H        North York
   
28         M3H        North York
   
29         M4H         East York
   
..         ...                ...   
73         M4R   Central Toronto
   
74         M5R   Central Toronto
   
75         M6R      West Toronto
   
76         M7R       Mississauga
   
77         M9R         Etobicoke
   
78         M1S       Scarborough
   
79         M4S   Central Toronto
   
80         M5S  Downtown Toronto
   
81         M6S      West Toronto
   
82         M1T       Scarborough
   
83         M4T   Central Toronto
   
84         M5T  Downtown Toronto
   
85         M1V       Scarborough
   
86         M4V   Central Toronto
   
87         M5V  Downtown Toronto
   
88         M8V         Etobicoke
   
89         M9V         Etobicoke
   
90         M1W       Scarborough
   
91         M4W  Downtown Toronto
   
92         M5W  Downtown Toronto
   
93         M8W         Etobicoke
   
94         M9W         Etobicoke
   
95         M1X       Scarborough
   
96         M4X  Downtown Toronto
   
97         M5X  Downtown Toronto
   
98         M8X         Etobicoke
   
99         M4Y  Downtown Toronto
   
100        M7Y      East Toronto
   
101        M8Y         Etobicoke
   
102        M8Z         Etobicoke
   

                                         Neighbourhood   Latitude  Longitude  
0                                           Parkwoods
  43.753259 -79.329656  
1                                    Victoria Village
  43.725882 -79.315572  
2                           Regent Park, Harbourfront
  43.654260 -79.360636  
3                    Lawrence Manor, Lawrence Heights
  43.718518 -79.464763  
4         Queen's Park, Ontario Provincial Government
  43.662301 -79.389494  
5             Islington Avenue, Humber Valley Village
  43.667856 -79.532242  
6                                      Malvern, Rouge
  43.806686 -79.194353  
7                                           Don Mills
  43.745906 -79.352188  
8                     Parkview Hill, Woodbine Gardens
  43.706397 -79.309937  
9                            Garden District, Ryerson
  43.657162 -79.378937  
10                                          Glencairn
  43.709577 -79.445073  
11   West Deane Park, Princess Gardens, Martin Grov...  43.650943 -79.554724  
12             Rouge Hill, Port Union, Highland Creek
  43.784535 -79.160497  
13                                          Don Mills
  43.725900 -79.340923  
14                                   Woodbine Heights
  43.695344 -79.318389  
15                                     St. James Town
  43.651494 -79.375418  
16                                 Humewood-Cedarvale
  43.693781 -79.428191  
17   Eringate, Bloordale Gardens, Old Burnhamthorpe...  43.643515 -79.577201  
18                  Guildwood, Morningside, West Hill
  43.763573 -79.188711  
19                                        The Beaches
  43.676357 -79.293031  
20                                        Berczy P

## Peer-graded Assignment: Segmenting and Clustering Neighborhoods in Toronto¶
# Problem - 3

In [55]:
t.head()

PostalCode            Borough                                 Neighbourhood  \
0        M3A        North York
                                    Parkwoods
   
1        M4A        North York
                             Victoria Village
   
2        M5A  Downtown Toronto
                    Regent Park, Harbourfront
   
3        M6A        North York
             Lawrence Manor, Lawrence Heights
   
4        M7A  Downtown Toronto
  Queen's Park, Ontario Provincial Government
   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

In [57]:
#Filter Borough as Toronto
df_toronto=t[t['Borough'].str.contains('Toronto', na=False)]
df_toronto.head()

PostalCode            Borough  \
2         M5A  Downtown Toronto
   
4         M7A  Downtown Toronto
   
9         M5B  Downtown Toronto
   
15        M5C  Downtown Toronto
   
19        M4E      East Toronto
   

                                   Neighbourhood   Latitude  Longitude  
2                     Regent Park, Harbourfront
  43.654260 -79.360636  
4   Queen's Park, Ontario Provincial Government
  43.662301 -79.389494  
9                      Garden District, Ryerson
  43.657162 -79.378937  
15                               St. James Town
  43.651494 -79.375418  
19                                  The Beaches
  43.676357 -79.293031

In [58]:
from geopy.geocoders import Nominatim
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="on_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [66]:
import folium

# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, Neighbourhood in zip(df_toronto['Latitude'], df_toronto['Longitude'], df_toronto['Borough'], df_toronto['Neighbourhood']):
    label = '{}, {}'.format(df_toronto, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [65]:
CLIENT_ID = 'F2ZD1SAQOCDK02QOICXTINKUAU4LAWSDPIOTYRVBXPILS2IT' # your Foursquare ID
CLIENT_SECRET = 'RHPGHO5KMKOWJOPMUNXOFAHKBHNP3KFOOFD1XLPF4335XX2F' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: F2ZD1SAQOCDK02QOICXTINKUAU4LAWSDPIOTYRVBXPILS2IT
CLIENT_SECRET:RHPGHO5KMKOWJOPMUNXOFAHKBHNP3KFOOFD1XLPF4335XX2F


The geograpical coordinate of Toronto are 43.6534817, -79.3839347 as derived from above code.

In [69]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    latitude, 
    longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=F2ZD1SAQOCDK02QOICXTINKUAU4LAWSDPIOTYRVBXPILS2IT&client_secret=RHPGHO5KMKOWJOPMUNXOFAHKBHNP3KFOOFD1XLPF4335XX2F&v=20180605&ll=43.6534817,-79.3839347&radius=500&limit=100'

In [71]:
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5f1b314c4d24b71550feee64'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Bay Street Corridor',
  'headerFullLocation': 'Bay Street Corridor, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 82,
  'suggestedBounds': {'ne': {'lat': 43.6579817045, 'lng': -79.37772678059432},
   'sw': {'lat': 43.6489816955, 'lng': -79.39014261940568}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '5227bb01498e17bf485e6202',
       'name': 'Downtown Toronto',
       'location': {'lat': 43.65323167517444,
        'lng': -79.38529600606677,
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.65323167517444,
          'lng'

In [72]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [73]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name        categories        lat        lng
0        Downtown Toronto      Neighborhood  43.653232 -79.385296
1  Nathan Phillips Square             Plaza  43.652270 -79.383516
2               Poke Guys        Poke Place  43.654895 -79.385052
3                 Japango  Sushi Restaurant  43.655268 -79.385165
4                  Indigo         Bookstore  43.653515 -79.380696

In [74]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

82 venues were returned by Foursquare.


In [83]:
df_toronto.head()

PostalCode            Borough  \
2         M5A  Downtown Toronto
   
4         M7A  Downtown Toronto
   
9         M5B  Downtown Toronto
   
15        M5C  Downtown Toronto
   
19        M4E      East Toronto
   

                                   Neighbourhood   Latitude  Longitude  
2                     Regent Park, Harbourfront
  43.654260 -79.360636  
4   Queen's Park, Ontario Provincial Government
  43.662301 -79.389494  
9                      Garden District, Ryerson
  43.657162 -79.378937  
15                               St. James Town
  43.651494 -79.375418  
19                                  The Beaches
  43.676357 -79.293031

In [85]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [87]:
df_toronto.columns

Index(['PostalCode', 'Borough', 'Neighbourhood', 'Latitude', 'Longitude'], dtype='object')

In [88]:
toronto_venues = getNearbyVenues(names=df_toronto['Neighbourhood'],
                                   latitudes=df_toronto['Latitude'],
                                   longitudes=df_toronto['Longitude']
                                  )

Regent Park, Harbourfront

Queen's Park, Ontario Provincial Government

Garden District, Ryerson

St. James Town

The Beaches

Berczy Park

Central Bay Street

Christie

Richmond, Adelaide, King

Dufferin, Dovercourt Village

Harbourfront East, Union Station, Toronto Islands

Little Portugal, Trinity

The Danforth West, Riverdale

Toronto Dominion Centre, Design Exchange

Brockton, Parkdale Village, Exhibition Place

India Bazaar, The Beaches West

Commerce Court, Victoria Hotel

Studio District

Lawrence Park

Roselawn

Davisville North

Forest Hill North & West, Forest Hill Road Park

High Park, The Junction South

North Toronto West,  Lawrence Park

The Annex, North Midtown, Yorkville

Parkdale, Roncesvalles

Davisville

University of Toronto, Harbord

Runnymede, Swansea

Moore Park, Summerhill East

Kensington Market, Chinatown, Grange Park

Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park

CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay

In [89]:
print(toronto_venues.shape)
toronto_venues.head()

(1637, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront
               43.65426              -79.360636   
1  Regent Park, Harbourfront
               43.65426              -79.360636   
2  Regent Park, Harbourfront
               43.65426              -79.360636   
3  Regent Park, Harbourfront
               43.65426              -79.360636   
4  Regent Park, Harbourfront
               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3     Body Blitz Spa East       43.654735       -79.359874   
4          Impact Kitchen       43.656369       -79.356980   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3                  Spa  
4           Restaurant

In [90]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            58   
                                                                       22   
Brockton, Parkdale Village, Exhibition Place                           16   
                                                                       18   
Business reply mail Processing Centre, South Ce...                     64   
CN Tower, King and Spadina, Railway Lands, Harb...                     16   
Central Bay Street                                                     75   
                                                                      100   
Christie                                                               35   
                                                                       10   
Church and Wellesley                                                   14   
                                                                      100   
Commerce Court, Victoria Hotel                                          4   
                                                                      100   
Davisville                                                            100   
                                                                       25   
Davisville North                                                       18   
                                                                       66   
Dufferin, Dovercourt Village                                            4   
                                                                       47   
First Canadian Place, Underground city                                  2   
                                                                       17   
Forest Hill North & West, Forest Hill Road Park                        14   
                                                                       34   
Garden District, Ryerson                                               45   
                                                                      100   
Harbourfront East, Union Station, Toronto Islands                       4   
                                                                        2   
High Park, The Junction South                                          36   
                                                                       85   
India Bazaar, The Beaches West                                         48   
                                                                       98   
Kensington Market, Chinatown, Grange Park                              41   
                                                                       16   
Lawrence Park                                                          22   
                                                                        5   
Little Portugal, Trinity                                               42   
                                                                      100   
Moore Park, Summerhill East                                            34   
                                                                            
North Toronto West,  Lawrence Park                                          
                                                                            
Parkdale, Roncesvalles                                                      
                                                                            
Queen's Park, Ontario Provincial Government                                 
                                                                            
Regent Park, Harbourfront                                                   
                                                                            
Richmond, Adelaide, King                                                    
                                                                            
Rosedale                                        

In [91]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 231 uniques categories.


In [92]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0                  0        0                   0             0   
1            0                  0        0                   0             0   
2            0                  0        0                   0             0   
3            0                  0        0                   0             0   
4            0                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  ...  Theater  Theme Restaurant  Toy / Game Store  Trail  \
0             0  ...        0                 0                 0      0   
1             0  ...        0                 0                 0      0   
2             0  ...        0                 0                 0      0   
3             0  ...        0                 0                 0      0   
4             0  ...        0                 0                 0      0   

   Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0              0                              0                 0   
1              0                              0                 0   
2              0                              0                 0   
3              0                              0                 0   
4              0                              0                 0   

   Vietnamese Restaurant  Wine Bar  Women's Store  
0                      0         0              0  
1                      0         0              0  
2                      0         0              0  
3                      0         0              0  
4                      0         0              0  

[5 rows x 231 columns]

In [93]:
toronto_onehot.shape

(1637, 231)

In [95]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                                        Berczy Park
     0.000000   
1       Brockton, Parkdale Village, Exhibition Place
     0.000000   
2   Business reply mail Processing Centre, South C...     0.000000   
3   CN Tower, King and Spadina, Railway Lands, Har...     0.000000   
4                                 Central Bay Street
     0.015625   
5                                           Christie
     0.000000   
6                               Church and Wellesley
     0.026667   
7                     Commerce Court, Victoria Hotel
     0.000000   
8                                         Davisville
     0.000000   
9                                   Davisville North
     0.000000   
10                      Dufferin, Dovercourt Village
     0.000000   
11            First Canadian Place, Underground city
     0.000000   
12   Forest Hill North & West, Forest Hill Road Park
     0.000000   
13                          Garden District, Ryerson
     0.000000   
14  Harbourfront East, Union Station, Toronto Isla...     0.000000   
15                     High Park, The Junction South
     0.000000   
16                    India Bazaar, The Beaches West
     0.000000   
17         Kensington Market, Chinatown, Grange Park
     0.000000   
18                                     Lawrence Park
     0.000000   
19                          Little Portugal, Trinity
     0.021277   
20                       Moore Park, Summerhill East
     0.000000   
21                North Toronto West,  Lawrence Park
     0.058824   
22                            Parkdale, Roncesvalles
     0.000000   
23       Queen's Park, Ontario Provincial Government
     0.029412   
24                         Regent Park, Harbourfront
     0.022222   
25                          Richmond, Adelaide, King
     0.000000   
26                                          Rosedale
     0.000000   
27                                          Roselawn
     0.000000   
28                                Runnymede, Swansea
     0.027778   
29                                    St. James Town
     0.000000   
30                       St. James Town, Cabbagetown
     0.000000   
31                                    Stn A PO Boxes
     0.010204   
32                                   Studio District
     0.024390   
33  Summerhill West, Rathnelly, South Hill, Forest...     0.000000   
34               The Annex, North Midtown, Yorkville
     0.000000   
35                                       The Beaches
     0.000000   
36                      The Danforth West, Riverdale
     0.023810   
37          Toronto Dominion Centre, Design Exchange
     0.000000   
38                    University of Toronto, Harbord
     0.029412   

    Afghan Restaurant   Airport  Airport Food Court  Airport Gate  \
0            0.000000  0.000000            0.000000      0.000000   
1            0.000000  0.000000            0.000000      0.000000   
2            0.000000  0.000000            0.000000      0.000000   
3            0.000000  0.055556            0.055556      0.055556   
4            0.000000  0.000000            0.000000      0.000000   
5            0.000000  0.000000            0.000000      0.000000   
6            0.013333  0.000000            0.000000      0.000000   
7            0.000000  0.000000            0.000000      0.000000   
8            0.000000  0.000000            0.000000      0.000000   
9            0.000000  0.000000            0.000000      0.000000   
10           0.000000  0.000000            0.000000      0.000000   
11           0.000000  0.000000            0.000000      0.000000   
12           0.000000  0.000000            0.000000      0.000000   
13           0.000000  0.000000            0.000000      0.000000   
14           0.000000  0.000000            0.000000      0.000000   
15           0.000000  0.000000            0.000000      0.000000   
16           0.000000  0.000000            0.000000      0.000000  

In [96]:
toronto_grouped.shape

(39, 231)

In [97]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park
----
          venue  freq
0   Coffee Shop  0.09
1          Café  0.03
2  Cocktail Bar  0.03
3        Bakery  0.03
4    Restaurant  0.03


----Brockton, Parkdale Village, Exhibition Place
----
                    venue  freq
0                    Café  0.14
1             Coffee Shop  0.09
2          Breakfast Spot  0.09
3  Furniture / Home Store  0.05
4                 Stadium  0.05


----Business reply mail Processing Centre, South Central Letter Processing Plant Toronto
----
                  venue  freq
0    Light Rail Station  0.12
1  Fast Food Restaurant  0.06
2         Garden Center  0.06
3           Pizza Place  0.06
4                  Park  0.06


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
----
              venue  freq
0   Airport Service  0.17
1    Airport Lounge  0.11
2  Airport Terminal  0.11
3   Harbor / Marina  0.06
4               Bar  0.06


----Central Bay Street
----
                ven

In [98]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [99]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                       Berczy Park
           Coffee Shop   
1      Brockton, Parkdale Village, Exhibition Place
                  Café   
2  Business reply mail Processing Centre, South C...    Light Rail Station   
3  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
4                                Central Bay Street
           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0              Beer Bar        Farmers Market          Cocktail Bar   
1        Breakfast Spot           Coffee Shop     Convenience Store   
2            Smoke Shop         Auto Workshop           Pizza Place   
3        Airport Lounge      Airport Terminal         Boat or Ferry   
4    Italian Restaurant                  Café        Sandwich Place   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                Bakery    Seafood Restaurant           Cheese Shop   
1          Climbing Gym               Stadium                   Bar   
2            Comic Shop         Burrito Place            Restaurant   
3       Harbor / Marina                   Bar   Rental Car Location   
4      Department Store                   Bar   Japanese Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                  Café            Restaurant              Irish Pub  
1    Italian Restaurant                Bakery             Restaurant  
2               Brewery            Skate Park                    Spa  
3      Sculpture Garden                 Plane            Coffee Shop  
4           Salad Place       Bubble Tea Shop           Burger Joint

In [115]:
from sklearn.cluster import KMeans

# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

In [118]:
neighborhoods_venues_sorted

Cluster Labels                                       Neighborhood  \
0                0                                       Berczy Park
   
1                0      Brockton, Parkdale Village, Exhibition Place
   
2                0  Business reply mail Processing Centre, South C...   
3                0  CN Tower, King and Spadina, Railway Lands, Har...   
4                0                                Central Bay Street
   
5                0                                          Christie
   
6                0                              Church and Wellesley
   
7                0                    Commerce Court, Victoria Hotel
   
8                0                                        Davisville
   
9                0                                  Davisville North
   
10               0                      Dufferin, Dovercourt Village
   
11               0            First Canadian Place, Underground city
   
12               3   Forest Hill North & West, Forest Hill Road Park
   
13               0                          Garden District, Ryerson
   
14               0  Harbourfront East, Union Station, Toronto Isla...   
15               0                     High Park, The Junction South
   
16               0                    India Bazaar, The Beaches West
   
17               0         Kensington Market, Chinatown, Grange Park
   
18               4                                     Lawrence Park
   
19               0                          Little Portugal, Trinity
   
20               1                       Moore Park, Summerhill East
   
21               0                North Toronto West,  Lawrence Park
   
22               0                            Parkdale, Roncesvalles
   
23               0       Queen's Park, Ontario Provincial Government
   
24               0                         Regent Park, Harbourfront
   
25               0                          Richmond, Adelaide, King
   
26               4                                          Rosedale
   
27               2                                          Roselawn
   
28               0                                Runnymede, Swansea
   
29               0                                    St. James Town
   
30               0                       St. James Town, Cabbagetown
   
31               0                                    Stn A PO Boxes
   
32               0                                   Studio District
   
33               0  Summerhill West, Rathnelly, South Hill, Forest...   
34               0               The Annex, North Midtown, Yorkville
   
35               0                                       The Beaches
   
36               0                      The Danforth West, Riverdale
   
37               0          Toronto Dominion Centre, Design Exchange
   
38               0                    University of Toronto, Harbord
   

   1st Most Common Venue          2nd Most Common Venue  \
0            Coffee Shop                       Beer Bar   
1                   Café                 Breakfast Spot   
2     Light Rail Station                     Smoke Shop   
3        Airport Service                 Airport Lounge   
4            Coffee Shop             Italian Restaurant   
5          Grocery Store                           Café   
6            Coffee Shop               Sushi Restaurant   
7            Coffee Shop                           Café   
8            Pizza Place                 Sandwich Place   
9                  Hotel                           Park   
10              Pharmacy                         Bakery   
11           Coffee Shop                           Café   
12         Jewelry Store                          Trail   
13        Clothing Store                    Coffee Shop   
14           Coffee Shop                       Aquarium   
15                  Café             Mexican Restaurant   
16             Pet Store              Fish & Chips Shop   
17                  C

In [109]:
df_toronto.rename(columns = {'Neighbourhood':'Neighborhood'}, inplace = True)
df_toronto.head()

/opt/conda/envs/Python36/lib/python3.6/site-packages/pandas/core/frame.py:4025: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  return super(DataFrame, self).rename(**kwargs)


PostalCode            Borough  \
2         M5A  Downtown Toronto
   
4         M7A  Downtown Toronto
   
9         M5B  Downtown Toronto
   
15        M5C  Downtown Toronto
   
19        M4E      East Toronto
   

                                    Neighborhood   Latitude  Longitude  
2                     Regent Park, Harbourfront
  43.654260 -79.360636  
4   Queen's Park, Ontario Provincial Government
  43.662301 -79.389494  
9                      Garden District, Ryerson
  43.657162 -79.378937  
15                               St. James Town
  43.651494 -79.375418  
19                                  The Beaches
  43.676357 -79.293031

In [132]:
neighborhoods_venues_sorted=neighborhoods_venues_sorted.drop(['Cluster Labels'], axis=1)

In [137]:
neighborhoods_venues_sorted.head()

Cluster Labels                                       Neighborhood  \
0               0                                       Berczy Park
   
1               0      Brockton, Parkdale Village, Exhibition Place
   
2               0  Business reply mail Processing Centre, South C...   
3               0  CN Tower, King and Spadina, Railway Lands, Har...   
4               0                                Central Bay Street
   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0           Coffee Shop              Beer Bar        Farmers Market   
1                  Café        Breakfast Spot           Coffee Shop   
2    Light Rail Station            Smoke Shop         Auto Workshop   
3       Airport Service        Airport Lounge      Airport Terminal   
4           Coffee Shop    Italian Restaurant                  Café   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0          Cocktail Bar                Bakery    Seafood Restaurant   
1     Convenience Store          Climbing Gym               Stadium   
2           Pizza Place            Comic Shop         Burrito Place   
3         Boat or Ferry       Harbor / Marina                   Bar   
4        Sandwich Place      Department Store                   Bar   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0           Cheese Shop                  Café            Restaurant   
1                   Bar    Italian Restaurant                Bakery   
2            Restaurant               Brewery            Skate Park   
3   Rental Car Location      Sculpture Garden                 Plane   
4   Japanese Restaurant           Salad Place       Bubble Tea Shop   

  10th Most Common Venue  
0              Irish Pub  
1             Restaurant  
2                    Spa  
3            Coffee Shop  
4           Burger Joint

In [139]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df_toronto

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood', how = 'right')

toronto_merged.head() # check the last columns!

PostalCode            Borough  \
2         M5A  Downtown Toronto
   
4         M7A  Downtown Toronto
   
9         M5B  Downtown Toronto
   
15        M5C  Downtown Toronto
   
19        M4E      East Toronto
   

                                    Neighborhood   Latitude  Longitude  \
2                     Regent Park, Harbourfront
  43.654260 -79.360636   
4   Queen's Park, Ontario Provincial Government
  43.662301 -79.389494   
9                      Garden District, Ryerson
  43.657162 -79.378937   
15                               St. James Town
  43.651494 -79.375418   
19                                  The Beaches
  43.676357 -79.293031   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
2                0           Coffee Shop                Bakery   
4                0           Coffee Shop                 Diner   
9                0        Clothing Store           Coffee Shop   
15               0                  Café           Coffee Shop   
19               0      Asian Restaurant     Health Food Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
2                   Park                   Pub               Theater   
4      College Cafeteria              Beer Bar         Smoothie Shop   
9        Bubble Tea Shop                  Café                 Hotel   
15        Clothing Store            Restaurant   American Restaurant   
19                   Pub                 Trail          Dance Studio   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
2                   Café        Breakfast Spot     Electronics Store   
4         Sandwich Place         Burrito Place                  Café   
9         Cosmetics Shop    Italian Restaurant   Japanese Restaurant   
15          Cocktail Bar        Cosmetics Shop              Creperie   
19         Deli / Bodega      Department Store          Dessert Shop   

   9th Most Common Venue 10th Most Common Venue  
2                  Hotel         Ice Cream Shop  
4                   Park     College Auditorium  
9         Lingerie Store              Bookstore  
15                   Gym         Farmers Market  
19                 Diner          Women's Store

In [141]:
import matplotlib.cm as cm
import matplotlib.colors as colors

# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

# Examine Clusters

In [143]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
83  Central Toronto
               1                   Gym   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
83          Tennis Court         Women's Store          Cupcake Shop   

          5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
83  Eastern European Restaurant   Dumpling Restaurant            Donut Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
83      Doner Restaurant               Dog Run    Distribution Center

In [144]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
62  Central Toronto
               2                  Pool   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
62                Garden         Women's Store      Cuban Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
62            Donut Shop      Doner Restaurant               Dog Run   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
62   Distribution Center        Discount Store                  Diner

In [145]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
68  Central Toronto
               3         Jewelry Store   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
68                 Trail    Mexican Restaurant      Sushi Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
68         Women's Store          Dance Studio   Dumpling Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
68            Donut Shop      Doner Restaurant                Dog Run